histogram/histogram.go

```diff
@@ -0,0 +1,54 @@
+// Package histogram provides primitives for working with histograms,
+// particularly towards Prometheus exporters.
+package histogram
+
+import (
+	"github.com/giantswarm/microerror"
+)
+
+type Config struct {
+	// BucketLimits is the upper limit of each bucket in the histogram.
+	// See https://godoc.org/github.com/prometheus/client_golang/prometheus#HistogramOpts.
+	BucketLimits []float64
+}
+
+// Histogram is a data structure suitable for providing the inputs to a Prometheus Histogram.
+// See https://godoc.org/github.com/prometheus/client_golang/prometheus#MustNewConstHistogram.
+type Histogram struct {
+	// Count is the number of entries added to the Histogram.
+	Count uint64
+	// Sum is the total sum of all entries added to the Histogram.
+	Sum float64
+	// Buckets is a map of upper bounds to cumulative counts of entries,
+	// excluding the +Inf bucket.
+	Buckets map[float64]uint64
+}
+
+func New(config Config) (*Histogram, error) {
+	if len(config.BucketLimits) == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Buckets must not be empty", config)
+	}
+
+	buckets := map[float64]uint64{}
+	for _, bucketLimit := range config.BucketLimits {
+		buckets[bucketLimit] = 0
```

Comment:
So what is `bucketLimit` supposed to be? We should have validation for this.

Reply:
In a nutshell, the upper limit for each bucket.

E.g: if you had buckets of 0, 10, 50, and wanted to add 25 to the histogram, it would be added to the first and second buckets.

I'm not certain what validation for these we could apply. You can have any list of numbers.

Reply:
So the bucket value defines the actual value of the metric and not a percentile when I understand you correctly. When this is the case there cannot be any useful validation. When the buckets determine the percentage then the validation should ensure only numbers from 1 to 100 work.

Reply:
It's just the values of the buckets, so agree that there's no useful validation.

Reply:
So what is when it is empty?